examples/pandas/split-apply-combine/my_functions.py

```diff
@@ -0,0 +1,179 @@
+import pandas as pd
+from pandas import DataFrame, Series
+
+from hamilton.function_modifiers import pipe, step, extract_fields, extract_columns, inject, source
+
+# ----------------------------------------------------------------------------------------------------------------------
+# Tax Rate & Credit rules
+# ----------------------------------------------------------------------------------------------------------------------
+
+tax_rates_rules = {
+    "Income < 50000": 0.15,  # < 50k: Tax rate is 15 %
+    "Income > 50000 and Income < 70000": 0.18,  # 50k to 70k: Tax rate is 18 %
+    "Income > 70000 and Income < 100000": 0.2,  # 70k to 100k: Tax rate is 20 %
+    "Income > 100000 and Income < 120000": 0.22,  # 100k to 120k: Tax rate is 22 %
+    "Income > 120000 and Income < 150000": 0.25,  # 120k to 150k: Tax rate is 25 %
+    "Income > 150000": 0.28,  # over 150k: Tax rate is 28 %
+}
+
+tax_credits_rules = {
+    "Children == 0": 0,  # 0 child: Tax credit 0 %
+    "Children == 1": 0.02,  # 1 child: Tax credit 2 %
+    "Children == 2": 0.04,  # 2 children: Tax credit 4 %
+    "Children == 3": 0.06,  # 3 children: Tax credit 6 %
+    "Children == 4": 0.08,  # 4 children: Tax credit 8 %
+    "Children > 4": 0.1,  # over 4 children: Tax credit 10 %
+}
+
+
+# ----------------------------------------------------------------------------------------------------------------------
+# Tax calculation functions
+# ----------------------------------------------------------------------------------------------------------------------
+
+def tax_rate(df: DataFrame, tax_rates: Dict[str, float]) -> DataFrame:
+    """
+    Add a series 'Tax Rate' to the DataFrame based on the tax_rates rules.
+    :param df: The DataFrame
+    :param tax_rates: Tax rates rules
+    :return: the DataFrame with the 'Tax Rate' Series
+    """
+    output = DataFrame()
+    for tax_rate_formula, tax_rate in tax_rates.items():
+        selected = df.query(tax_rate_formula)
+        if selected.empty:
+            continue
+        tmp = DataFrame({"Tax Rate": tax_rate}, index=selected.index)
+        output = pd.concat([output, tmp], axis=0)
+    df["Tax Rate"] = output["Tax Rate"]
+    return df
+
+
+def tax_credit(df: DataFrame, tax_credits: Dict[str, float]) -> DataFrame:
+    """
+    Add a series 'Tax Credit' to the DataFrame based on the tax_credits rules.
+    :param df: The DataFrame
+    :param tax_credits: Tax credits rules
+    :return: the DataFrame with the 'Tax Credit' Series
+    """
+    output = DataFrame()
+    for tax_credit_formula, tax_credit in tax_credits.items():
+        selected = df.query(tax_credit_formula)
+        if selected.empty:
+            continue
+        tmp = DataFrame({"Tax Credit": tax_credit}, index=selected.index)
+        output = pd.concat([output, tmp], axis=0)
+    df["Tax Credit"] = output["Tax Credit"]
+    return df
+
+
+@inject(income=source("Income"), tax_rate=source("Tax Rate"), tax_credit=source("Tax Credit"))
```

Comment:
```suggestion
# we use inject here because we have spaces in the names of columns. If column names are valid python variable names we wouldn't need this.
@inject(income=source("Income"), tax_rate=source("Tax Rate"), tax_credit=source("Tax Credit"))
```